sputnikdao2/src/upgrade.rs

```diff
@@ -0,0 +1,134 @@
+//! Logic to upgrade Sputnik contracts.
+
+use near_sdk::Gas;
+
+use crate::*;
+
+const FACTORY_KEY: &[u8; 7] = b"FACTORY";
+const DEFAULT_FACTORY_ID: &str = "sputnik-dao.near";
+const ERR_MUST_BE_SELF_OR_FACTORY: &str = "ERR_MUST_BE_SELF_OR_FACTORY";
+const SELF_MIGRATE_METHOD_NAME: &[u8; 7] = b"migrate";
+const UPDATE_GAS_LEFTOVER: Gas = Gas(5_000_000_000_000);
+const NO_DEPOSIT: Balance = 0;
+
+/// Gas for upgrading this contract on promise creation + deploying new contract.
+pub const GAS_FOR_UPGRADE_SELF_DEPLOY: Gas = Gas(30_000_000_000_000);
+
+pub const GAS_FOR_UPGRADE_REMOTE_DEPLOY: Gas = Gas(10_000_000_000_000);
+
+/// Info about factory that deployed this contract and if auto-update is allowed.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Clone, Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub struct FactoryInfo {
+    pub factory_id: AccountId,
+    pub auto_update: bool,
+}
+
+/// Fetches factory info from the storage.
+/// By design not using contract STATE to allow for upgrade of stuck contracts from factory.
+pub(crate) fn internal_get_factory_info() -> FactoryInfo {
+    env::storage_read(FACTORY_KEY)
+        .map(|value| FactoryInfo::try_from_slice(&value).expect("INTERNAL_FAIL"))
+        .unwrap_or_else(|| FactoryInfo {
+            factory_id: AccountId::new_unchecked(DEFAULT_FACTORY_ID.to_string()),
+            auto_update: true,
+        })
+}
+
+pub(crate) fn internal_set_factory_info(factory_info: &FactoryInfo) {
+    env::storage_write(
+        FACTORY_KEY,
+        &factory_info.try_to_vec().expect("INTERNAL_FAIL"),
+    );
+}
+
+/// Function that receives new contract, updates and calls migration.
+/// Two options who call it:
+///  - current account, in case of fetching contract code from factory;
+///  - factory, if this contract allows to factory-update;
+#[no_mangle]
+pub fn update() {
+    env::setup_panic_hook();
+    let factory_info = internal_get_factory_info();
+    let current_id = env::current_account_id();
+    assert!(
+        env::predecessor_account_id() == current_id
+            || (env::predecessor_account_id() == factory_info.factory_id
+                && factory_info.auto_update),
+        "{}",
+        ERR_MUST_BE_SELF_OR_FACTORY
+    );
+    unsafe {
+        // Load code into register 0 result from the promise.
+        sys::promise_result(0, 0);
```

Comment:
@ilblackdragon , I've just tested your entire code and everything runs smoothly except for the `upgrade` functionality.

I kept getting `Accessed invalid promise result index: 0` so I started debugging it. Apparently, the error gets generated from this line.

I've looked into it and something doesn't add up. You're trying to get the promise result from the previous call, but there is no call to get the result from. The only call is the one that you're actually in right now.

If you read the description of `promise_result` function, it says:
```
    /// If the current function is invoked by a callback we can access the execution results of the
    /// promises that caused the callback. This function returns the result in blob format and
    /// places it into the register.
```

The problem is that you don't call `promise_result` from the callback here. You're calling it from the call itself, so there's no result available.

My suggestion is to eliminate `promise_result` and get the contract code from the input context. Like you do on `store_contract`.

Reply:
```suggestion
        sys::input(0);
```

Reply:
```suggestion
        sys::input(0);
```

Reply:
Good point. Specifically it should be sys::input if the call is from the factory and promise_result if it's a callback